presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -2722,7 +2722,8 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
                 tablePartitioning,
                 Optional.empty(),
                 discretePredicates,
-                ImmutableList.of());
+                ImmutableList.of(),
+                Optional.of(hiveLayoutHandle.getRemainingPredicate()));
```

Comment:
`getTableLayout` and `HiveFilterPushdown` are two ways of pushing down predicates. The previous one is the legacy one that can only push down tuple domains. The latter can push down arbitrary filters (or agg, project, etc). Remaining predicates is the `getTableLayout` one that returns back the optimizer to tell it "what is remaining in the filter after the tuple domain has been pushed down". But it doesn't really necessary mean it will be pushed down later by `HiveFilterPushdown`.

So, instead of calling it `pushdownPredicate`, we can just call it `remainingPredicate` and comment on it that it is only the ones after `getTableLayout`.

Reply:
Thanks for the background. IIUC, the dynamicFilterExpression below is the one which returned to the optimizer after the filter pushdown. in visitFilter, we create a new Filter with these expressions to keep them. Looks like the remainingExpression below is the predicate pushed down to the scan node. And these two are mutually exclusive, right? Do you have a code pointer where part of the remainingExpression is not pushed down by HiveFilterPushdown?

https://github.com/prestodb/presto/blob/d30100cb7cb9cb9890fe04447f82c4026cdf8a9b/presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java#L261-L262

Reply:
`dynamicFilterExpression` is a complete different concept. Check #15077 for details. It is used for dynamic join filtering by pushing down the build side digest to probe side for efficient pruning. So, let's don't worry that code path.

So `remainingExpression` is only associated with `getTableLayout`. `getTableLayout` is always called no matter what. So `remainingExpression` should always be present. Suppose our original filter is `partition_col_1 > 5 and foo(col2) < 3`. Then the tuple domain of `partition_col_1 > 5` will be pushed down and the `remainingExpression` will be `foo(col2) < 3`.

`HiveFilterPushdown` is a later concept we introduced years after `getTableLayout` to handle arbitrary filter pushdown beyond tuple domains. In order to make `getTableLayout` still working yet being backward compatible, we sort made `HiveFilterPushdown` to take care of whatever the filter is. So, in this case, `remainingExpression` it is. But it is not always guaranteed. For example, `HiveFilterPushdown` is enabled only when (1) `hive.pushdown_filter_enabled` is true, (2) the source table is a Hive table, and (3) the file format is ORC or DWRF. So there will be cases that filters won't be pushed down for MySQL tables or Hive tables with parquet formats.

I think it should be safe to use `remainingExpression` as an indicator but maybe not call it `pushdownPredicate` because it might not be pushed down.

Reply:
Thanks for the elaboration. Updated the names.